### PR TITLE
check pending reads when quorum becomes small

### LINKED ISSUE #345: Quorum change should check read index
When quorum changes due to configuration change or quorum function change, read index should be checked eagerly. Otherwise read index will not be advanced until next heartbeat.

### PR DESCRIPTION
**Pull request by hicqu**
Close #345 .

Signed-off-by: qupeng <qupeng@pingcap.com>

### CODE REVIEW

File: src/raft.rs [hunk 1]
```diff
@@ -2325,4 +2305,21 @@ impl<T: Storage> Raft<T> {
         m.request_snapshot = self.pending_request_snapshot;
         self.send(m);
     }
+
+    fn response_ready_read(&mut self, mut req: Message, index: u64) -> Option<Message> {
```
Comment (BusyJay) [inline src/raft.rs:2309]: s/response_ready_read/handle_ready_read_index/

File: src/raft.rs [hunk 2]
```diff
@@ -2214,6 +2175,25 @@ impl<T: Storage> Raft<T> {
         if self.maybe_commit() {
             self.bcast_append();
         }
+
+        // The quorum size is now smaller, consider to response some read requests.
+        // If there is only one peer, all pending read requests must be responsed.
+        if let Some(ctx) = self.read_only.last_pending_request_ctx() {
```
Comment (BusyJay) [inline src/raft.rs:2181]: What if some reads in the front of the queue are ready to be processed?

File: harness/tests/integration_cases/test_raft.rs [hunk 3]
```diff
@@ -4920,3 +4920,36 @@ fn test_group_commit_consistent() {
         }
     }
 }
+
+#[test]
+fn test_read_when_quorum_becomes_small() {
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft.rs:4925]: Please add comment what it is testing.

File: src/raft.rs (hunk 2 above)
Comment (hicqu) [reply src/raft.rs:2181]: Without iterating the pending queue, they can only become ready when more `heartbeat responses` are received. Do you have any solution to handle them without iterating?

**Pull request merged**